File: JavaPools/PoolsLibrary/src/main/java/ru/urfu/javapools/poolslibrary/pool/wrappers/PWSingleUseEnforcingWrapper.java
package ru.urfu.javapools.poolslibrary.pool.wrappers;

import ru.urfu.javapools.poolslibrary.controller.InvalidPoolOperationException;
import ru.urfu.javapools.poolslibrary.controller.PoolObjectReleasingInterruptedException;
import ru.urfu.javapools.poolslibrary.pool.IPool;
import ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item.PoolItem;

public class PWSingleUseEnforcingWrapper<TK,TV> extends PWBaseWrapper<TK,TV> {

	public PWSingleUseEnforcingWrapper(IPool<TK, TV> basePool) {
		super(basePool);
	}
	
	@Override
	public void release(TK key, TV objectToRelease)
		throws InvalidPoolOperationException, PoolObjectReleasingInterruptedException {
		
		PoolItem<TK,TV> poolItem = getKeyToPoolItem().get(key);
		poolItem.release(objectToRelease);
		
		_basePool.release(key, objectToRelease);
	}
}

File: CSharpPools/PoolsLibrary/Pool/Wrappers/PWSingleUseEnforcingWrapper.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PoolsLibrary.ObjectActions;
using PoolsLibrary.Pool.BasicFunctionality.Item;

namespace PoolsLibrary.Pool.Wrappers
{
    public class PWSingleUseEnforcingWrapper<TK,TV> : IInternalPool<TK,TV>
    {
        ConcurrentDictionary<TK, PoolItem<TK, TV>> IInternalPool<TK, TV>.KeyToPoolItem
        {
            get { return _basePool.KeyToPoolItem; }
        }

        private readonly IInternalPool<TK, TV> _basePool;

        public PWSingleUseEnforcingWrapper(IPool<TK, TV> basePool)
        {
            _basePool = basePool as IInternalPool<TK,TV>;
        }

        public bool TryObtain(TK key, out TV outPoolObject, Func<TK, TV> createDelegateIfNoObjectIsAvailable)
        {
            return _basePool.TryObtain(key, out outPoolObject, createDelegateIfNoObjectIsAvailable);
        }

        public void Release(TK key, TV objectToRelease)
        {
            PoolItem<TK, TV> outPoolItem;
            ((IInternalPool<TK,TV>)this).KeyToPoolItem.TryGetValue(key, out outPoolItem);
            outPoolItem.Release(objectToRelease);

            _basePool.Release(key, objectToRelease);
        }

        public void Dispose()
        {
            _basePool.Dispose();
        }
    }
}
